src/include/storage/index/index.h

```diff
@@ -57,19 +45,23 @@ class Index {
 
   /**
    * Creates a new index wrapper.
-   * @param constraint_type type of index
    * @param metadata index description
    */
-  Index(const ConstraintType constraint_type, IndexMetadata metadata)
-      : constraint_type_{constraint_type}, metadata_(std::move(metadata)) {}
+  explicit Index(IndexMetadata metadata) : metadata_(std::move(metadata)) {}
 
  public:
   virtual ~Index() = default;
 
+  /**
+   * @return type of the index. Note that this is the physical type, not extracted from the underlying schema or other
+   * catalog metadata. This is mostly used for debugging purposes.
+   */
+  virtual IndexType Type() const = 0;
+
   /**
    * Invoke garbage collection on the index. For some underlying index types this may be a no-op.
    */
-  virtual void PerformGarbageCollection() = 0;
+  virtual void PerformGarbageCollection() {}
```

Comment:
We can still keep it as a pure virtual function?

Reply:
I guess it depends on which we believe is more common for future index types. I actually believe we won't have to invoke `PerformGarbageCollection` on future indexes, so I leave it a no-op in the base class and leave it up to the derived class to override if needed. If, in the future, we find that most index types do need it, we can change it back to pure virtual and have the hash_index override it as a no-op.